monadic-rx/js/src/test/scala/mhtml/RxJsTests.scala

```diff
@@ -0,0 +1,59 @@
+package mhtml
+
+import scala.scalajs.js.timers._
+import org.scalatest.FunSuite
+
+class RxJsTests extends FunSuite with RxTestHelpers {
+
+  test("merge registrations are tail-shareable") {
+
+    val delay: Double = 25 // ms
+
+    val rx1: Var[Int] = Var(0)
+    val rx2: Var[Int] = Var(1)
+    val merged: Rx[Int] = rx1.merge(rx2)
+    var rx1List: List[Int] = Nil
+    var rx2List: List[Int] = Nil
+    var merged1List: List[Int] = Nil
+    var merged2List: List[Int] = Nil
+    val cc1 = rx1.impure.run(n => rx1List = rx1List :+ n)
+    val cc2 = rx2.impure.run(n => rx2List = rx2List :+ n)
+    var ccm1: Cancelable = Cancelable.empty
+    var ccm2: Cancelable = Cancelable.empty
+    setTimeout(1 * delay){
```

Comment:
I'm confused, if you are going to execute these sequentially, what's the point of `setTimeout`? Couldn't you simply move all these statements to the two level and get the same result?

Reply:
Apparently not, I had to modify the results of the asserts to pass as follows (which now appears to indicate *all-shareable():

```scala
      rx1 := 0
      rx2 := 1

      ccm1 = merged.impure.run(n => merged1List = merged1List :+ n)

      rx1 := 8

      rx2 := 4

      ccm2 = merged.impure.run(n => merged2List = merged2List :+ n)

      rx2 := 3

      rx1 := 3

      assert(rx1List == List(0, 0, 8, 3))
      assert(rx2List == List(1, 1, 4, 3))
      assert(merged1List == List(0, 1, 8, 4, 3, 3))
      assert(merged2List == List(8, 4, 3, 3))
```

Reply:
It's really surprising than the two are different...

I just realized than I forgot something about merge, it emits *both* initial values in order, which is why `val out = rx1.dropIf(_ % 2 == 0)(-1).merge(rx2.dropIf(_ % 2 == 0)(-1))` is worse than tail-sharable. Maybe we should change that such that merge only emits the initial value from the left stream, it would at least make things compose more nicely.

The implementation is [here](https://github.com/OlivierBlanvillain/monadic-html/blob/master/monadic-rx/shared/src/main/scala/mhtml/rx.scala#L227-L230). Somehow the effect resisted to the `other` stream should discard the first element. Something like this maybe:

```scala
case Merge(self, other) =>
   val c1 = run(self)(effect)
   var discard = true
   val c2 = run(other) { x =>
      if (discard) discard = false
      else effect(x)
   }
   Cancelable { () => c1.cancel; c2.cancel }
```

Reply:
1. I agree it is worth considering the sequential issue a bit more - may look at it more tomorrow if not today.

2. >  it emits both initial values in a fixed order (left then right)

This didn't seem to be the case in the added merge example? `merge1List` in the sequential example starts as '0, 8' - if it was as you describe, wouldn't it start as '0, 1'?

3. If the semantics of `merge` were changed as proposed, I suppose it would always be simple enough to initialize any "right-side rx" with a throw-away value, so this should be ok - the user would just need to be aware of the potential gotcha.

Reply:
(of course, it did start as '0, 1' in the non-`setTimeout` version, hmm)

Reply:
> This didn't seem to be the case in the added merge example?

The example is wrong

Reply:
Sorry - could you explain if you mean it is wrong in that it is a bug in the core Rx code, or just that it doesn't meet the current expectations of emitting both left and right "initial" values? Because I believe (unless I typo'd - quite possible) the added example exactly corresponds to the test I added that uses `setTimeout`.

Reply:
Sorry it wasn't clean :smile:. The diagram I draw for merge is wrong according to my understanding of how merge works, I don't understand why the test case you wrote using `setTimeout` behaves this way.